lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+			}
+			else
+			{
+				DebugErrorHandler::log("Ignoring corrupted socket {}.", s);
+			}
+
+			cell = cell->cell_next();
+		}
+	}
+
+	DebugErrorHandler::log("Resetting the sealed sockets list.");
+	sealedSockets.reset();
+
+	// Upgrade the two critical section locks for destruction
+	// TODO document what this will do
+	DebugErrorHandler::log("Upgrading critical sections for destruction.");
+	flaglock_upgrade_for_destruction(&__CriticalSectionFlagLock.lock);
+	flaglock_upgrade_for_destruction(&__SuspendFlagLock.lock);
+
+	// Upgrade the message queue lock for destruction
+	DebugErrorHandler::log("Upgrading the message queue for destruction.");
+	// TODO document what this will do
+	auto *queueHandle = &xNetworkEventQueue->handle;
+	if (int err = queue_destroy(MALLOC_CAPABILITY, queueHandle); err != 0)
+	{
+		DebugErrorHandler::log(
+		  "Failed to upgrade the message queue for destruction (error {}).",
+		  err);
+	}
+
+	// Wait for all user threads to exit.
+	// TODO Here, we can experiment with `switcher_interrupt_thread` to get
+	// threads to die faster.
+	DebugErrorHandler::log("Waiting for all threads to exit.");
+	while (userThreadCount.load() != 0)
+	{
+		DebugErrorHandler::log("Waiting for {} user thread(s) to terminate.",
+		                       userThreadCount.load());
+
+		// Threads may also be waiting on the allocator in an
+		// out-of-memory situation. Do a `heap_free_all` to unblock
+		// them. We must do this in the loop body in case threads
+		// re-enter OOM multiple times.
+		//
+		// We will do another free at the end of the reset to ensure
+		// that everything is cleaned up in case threads allocate
+		// memory again before terminating.
+		free_compartment_memory();
+
+		Timeout t{1};
+		thread_sleep(&t);
+	}
+
+	// Wait for the IP thread to reset (unless this error handler is
+	// running from the IP thread).
+	if (isUserThread)
+	{
+		DebugErrorHandler::log("Waiting for the IP thread to reset.");
+		// We will only manage to lock this when the IP thread releases
+		// the lock, which will happen when it re-enters its
+		// initialization phase.
+		flaglock_lock(&ipThreadLockState);
+		// Release the lock as we want the IP thread to grab it again
+		// when we unleash it.
+		flaglock_unlock(&ipThreadLockState);
+	}
+
+	// At this point all user threads have exited the TCP/IP stack
+	// compartment and the network thread context has been reinstalled.
+	DebugErrorHandler::Assert(userThreadCount.load() == 0,
+	                          "All user threads should be terminated by now.");
+
+	// Free heap memory.  We must do this *again*, because threads may have
+	// allocated memory since the previous calls to `heap_free_all`.
+	DebugErrorHandler::log("Freeing heap memory.");
+	free_compartment_memory();
+
+	// Update the socket epoch. We want to do this after all threads have
+	// terminated in case some threads were allocating new sockets during
+	// the restart.
+	currentSocketEpoch++;
```

Comment:
If we do this at the start and end of the epoch, threads can simply fail if the epoch number is odd (or if it changes during the socket creation).

Reply:
I initially decided not to do it, mainly because we still need the `restartState` (which is currently used to detect a restart) because this is our state machine and it's more complex that just "is restart" or "is not restart".

As we discussed separately however, we could merge the `restartState` and the epoch into a single atomic.

We leave this for future works, I added a comment.